### ISSUE TITLE
[RFC][KV Offload]: Align sliding-window restore coverage with MTP-retained history

### ISSUE THREAD

**Issue by jacklin78911-collab** [vllm-project/vllm#56701]
### Motivation

Native KV offload can report a sliding-window prefix hit and then try to load an older chunk that was never included in that hit. I reproduced both `Chunk ... not found in cache` and `Chunk ... is not ready for reading` through the real Python offload scheduler, CPU manager, and GPU block allocator classes, without concurrent eviction or any GPU kernel execution.

The trigger is `SlidingWindowSpec.extra_retained_tokens > 0`. `get_kv_cache_configs()` sets this to `num_speculative_tokens - 1` for multi-module MTP. That field extends the history retained by the block manager, but offload lookup and the producer's reachable-store mask still use the attention window without that extension.

This RFC asks whether the offload contract should cover the entire retained interval. My proposed first fix is to make lookup and store coverage match the existing allocation contract. The alternative—allocating only the attention-required interval on an external hit—needs a separate proof that later MTP replay and caching cannot read or publish the omitted history.

Source inspected and exercised: upstream main at [`fa1b3b1922ccb2d7a597925765c4bec7f358d748`](https://github.com/vllm-project/vllm/commit/fa1b3b1922ccb2d7a597925765c4bec7f358d748).

### Reproduced failure

A small hybrid layout is sufficient:

- Full-attention group: 64 tokens per block.
- Sliding-window group: 16 tokens per block, attention window 32, extra retained history 3.
- One block per offload chunk, 16 tokens per hash, 65-token prompt, zero local cache hits.
- CPU cache contains the full-attention prefix and SWA chunks 2 and 3. SWA chunk 1 is independently missing, pending, or ready.

At the 64-token hit boundary, attention needs history starting at token `64 - 32 + 1 = 33`, within SWA chunk 2. The actual allocator retains history starting at `64 - 32 + 1 - 3 = 30`, within chunk 1.

| Extra history | Chunk 1 | Reported hit | Allocated SWA offsets | Load result |
|---|---|---:|---|---|
| 0 | missing / pending / ready | 64 | 2, 3 | succeeds |
| 3 | missing | 64 | 1, 2, 3 | `not found in cache` assertion |
| 3 | pending | 64 | 1, 2, 3 | `not ready for reading` assertion |
| 3 | ready | 64 | 1, 2, 3 | succeeds |

The missing chunk does not require an eviction race. For this hybrid layout, the real `_reachable_store_block_mask()` returns `[False, False, True, True]` at the 64-token boundary even with extra history enabled: the producer's filter itself omits chunk 1.

This is a bookkeeping reproduction, not a full model-serving reproduction. The prompt hashes and config objects are synthetic; `OffloadingConnectorScheduler`, `CPUOffloadingManager`, `BlockPool`, `FullAttentionManager`, and `SlidingWindowManager` are the actual implementations. No scheduler or allocator function is mocked. The target SWA descriptor is populated directly to isolate the retention contract; the probe does not instantiate an MTP drafter.

### Code path

1. [`get_kv_cache_configs()`](https://github.com/vllm-project/vllm/blob/fa1b3b1922ccb2d7a597925765c4bec7f358d748/vllm/v1/core/kv_cache_utils.py) tags sliding-window specs for multi-module MTP.
2. [`SlidingWindowManager.get_num_skipped_tokens()`](https://github.com/vllm-project/vllm/blob/fa1b3b1922ccb2d7a597925765c4bec7f358d748/vllm/v1/core/single_type_kv_cache_manager.py) subtracts `extra_retained_tokens`. Initial local-block adoption pads fewer null slots; external allocation therefore allocates the older retained block too.
3. [`get_sliding_window_size_in_chunks()` / `load_window_size_in_chunks()`](https://github.com/vllm-project/vllm/blob/fa1b3b1922ccb2d7a597925765c4bec7f358d748/vllm/distributed/kv_transfer/kv_connector/v1/offloading/scheduler.py) omit that extension. Lookup reports 64 matched tokens without checking chunk 1.
4. `update_state_after_alloc()` starts from the first fresh non-null block, including chunk 1. [`CPUOffloadingManager.prepare_load()`](https://github.com/vllm-project/vllm/blob/fa1b3b1922ccb2d7a597925765c4bec7f358d748/vllm/v1/kv_offload/cpu/manager.py) then correctly rejects the unavailable chunk.
5. `SlidingWindowManager.reachable_block_mask()`, used by offload store filtering, also derives coverage from `sliding_window` alone.

### Proposed change

Keep the current allocator and MTP retention semantics. Make native offload coverage include every chunk that those semantics cause an external restore to read:

1. Derive offload lookup coverage from the retained interval, including partial-chunk alignment. Apply the same interval to the in-flight-load conflict check and relevant recency touches.
2. Extend the producer's offload store mask to supply that interval. Changing the lookup window alone would turn affected restores into misses because the producer can still omit the required chunk.
3. Keep the current outcomes: a missing required chunk shortens/rejects the hit; a pending required chunk defers lookup; a complete hit can be allocated and loaded. Preserve the existing `prepare_load` assertions.
4. Keep the actual model attention window, wire format, cache keys, and transfer ownership unchanged. No new user option is needed for an internal correctness contract.

The invariant is: **every chunk consumed by an admitted restore must be covered by the confirmed lookup, and every required chunk must be reachable through the corresponding producer store policy.**

A diagnostic baseline expands only the offload coverage descriptor from window 32 to 35 while preserving the real allocator's original `window=32, extra=3` spec. With this expansion, the mask becomes `[False, True, True, True]`; a missing chunk produces zero hits, a pending chunk defers, and a ready chunk restores 64 tokens successfully. All six `extra=0` controls are unchanged across the two modes. The script below checks all 12 combinations.

This descriptor substitution is an experiment, not the proposed production implementation. Production code should express offload coverage explicitly rather than changing a model's attention window. Wider coverage also has a cost: this example stores/loads three SWA chunks instead of two. It needs measurement, not an assumption that extra history is free.

### Validation before a PR

- Protect the confirmed missing/pending/ready behavior and zero-extra controls in the existing offload scheduler suite.
- Exercise a producer-to-consumer replay using actual store selection, CPU completion, lookup, and allocation; check that demanded keys were offered for storage.
- Cover chunk factors greater than one, unaligned hybrid hit boundaries, EAGLE tail handling, and in-flight loads with the same coverage invariant.
- Validate real multi-module MTP outputs and acceptance, plus cache bytes/hit rate and end-to-end latency on a prefix-reusing workload. These model/performance results have not been run and are not claimed here.

### Related work and duplicate check

Searched open PRs and issues for `extra_retained_tokens`, offload + retained history, multi-module + offload, and offload window/load coverage on 2026-09-13.

- #55818 addresses sliding-window allocation scaling and preserves extra retention; it does not align offload lookup/store coverage.
- #52287 addresses store windows moved by EAGLE's volatile-tail pop. This reproduction has no EAGLE pop.
- #56486 addresses a pending flag left behind by an invalid suffix. This reproduction has a complete reported hit and a later load failure.
- #54914 reports a similar load assertion under DSpark. I have not reproduced that deployment or established this mechanism as its cause; this RFC does not claim to resolve it.

### Feedback period

One week. The main decision is whether native offload should restore the entire retained interval, as proposed here, or whether initial external-hit allocation should distinguish retention-only history from the history required to resume. The latter must also preserve MTP replay and cache-publication correctness.

### Reproducer and environment

The 12-case diagnostic was executed with Python 3.12.13, PyTorch 2.11.0+cu130, the pinned Python source above, and existing precompiled extensions in WSL Ubuntu 24.04. No GPU kernels or model inference were executed. This is not validation on the repository's current PyTorch 2.13 dependency matrix.

Save the following as `retained_window_probe.py` in an installed checkout and run `.venv/bin/python retained_window_probe.py`. It writes a JSON result beside the script and checks the expected outcomes, including the two assertions in the unmodified path.

<details>
<summary>Executable 12-case diagnostic</summary>

```python
"""Exercise real offload lookup/allocation with MTP's retained SWA history."""

import json
from dataclasses import replace
from pathlib import Path
from types import SimpleNamespace
from unittest.mock import MagicMock

import torch

from vllm.distributed.kv_transfer.kv_connector.v1.offloading.scheduler import (
    OffloadingConnectorScheduler,
)
from vllm.v1.core.block_pool import BlockPool
from vllm.v1.core.kv_cache_manager import KVCacheBlocks
from vllm.v1.core.kv_cache_utils import BlockHash
from vllm.v1.core.single_type_kv_cache_manager import (
    FullAttentionManager,
    SlidingWindowManager,
)
from vllm.v1.kv_cache_interface import (
    FullAttentionSpec,
    KVCacheConfig,
    KVCacheGroupSpec,
    SlidingWindowSpec,
)
from vllm.v1.kv_offload.base import OffloadingKVEventsConfig
from vllm.v1.kv_offload.cpu.manager import CPUOffloadingManager


def probe(extra: int, older_state: str, expanded_coverage: bool = False):
    kv_spec = SlidingWindowSpec(
        block_size=16,
        num_kv_heads=1,
        head_size=1,
        dtype=torch.float32,
        sliding_window=32,
        extra_retained_tokens=extra,
    )
    full_spec = FullAttentionSpec(
        block_size=64,
        num_kv_heads=1,
        head_size=1,
        dtype=torch.float32,
    )
    groups = [KVCacheGroupSpec(["full"], full_spec), KVCacheGroupSpec(["swa"], kv_spec)]
    manager = CPUOffloadingManager(num_chunks=8)
    spec = SimpleNamespace(
        config=SimpleNamespace(
            groups=tuple(
                SimpleNamespace(group_id=i, tokens_per_block=g.kv_cache_spec.block_size)
                for i, g in enumerate(groups)
            )
        ),
        tokens_per_block=(64, 16),
        tokens_per_hash=16,
        blocks_per_chunk=1,
        offload_prompt_only=True,
        kv_events_config=OffloadingKVEventsConfig(
            enable_kv_cache_events=False, self_describing_kv_events=False
        ),
        get_manager=lambda: manager,
    )
    config = SimpleNamespace(
        speculative_config=None,
        cache_config=SimpleNamespace(
            enable_prefix_caching=True, prefix_cache_retention_interval=None
        ),
        parallel_config=SimpleNamespace(world_size=1, decode_context_parallel_size=1),
    )
    sched = OffloadingConnectorScheduler(
        spec,
        config,
        KVCacheConfig(num_blocks=32, kv_cache_tensors=[], kv_cache_groups=groups),
    )
    if expanded_coverage:
        # Diagnostic only: expand the offload coverage descriptor, preserving the
        # real GPU allocator's spec. This is not a production model patch.
        coverage_spec = replace(
            kv_spec, sliding_window=32 + extra, extra_retained_tokens=0
        )
        group = sched.config.kv_group_configs[1]._replace(
            sliding_window_size_in_chunks=(32 + extra + 15) // 16,
            kv_cache_spec=coverage_spec,
        )
        sched.config = sched.config._replace(
            kv_group_configs=(sched.config.kv_group_configs[0], group)
        )
    req = MagicMock()
    req.request_id = "replay"
    req.kv_transfer_params = None
    req.num_tokens = req.num_prompt_tokens = 65
    req.block_hashes = [BlockHash(f"block{i}".encode()) for i in range(4)]
    req.skip_reading_prefix_cache = False
    sched.on_new_request(req)
    state = sched._req_status[req.request_id]
    state.update_offload_keys()
    full_keys, keys = (g.offload_keys for g in state.group_states)
    ready = full_keys[:1] + keys[1 if older_state == "ready" else 2 :]
    store = manager.prepare_store(ready, state.req_context)
    assert store is not None
    manager.complete_store(store.keys_to_store, state.req_context)
    if older_state == "pending":
        assert manager.prepare_store([keys[1]], state.req_context) is not None
    external, async_load = sched.get_num_new_matched_tokens(req, 0)
    pool = BlockPool(32, enable_caching=True, hash_block_size=16)
    managers = [
        cls(
            kv_cache_spec=g.kv_cache_spec,
            block_pool=pool,
            enable_caching=True,
            kv_cache_group_id=i,
            scheduler_block_size=64,
        )
        for i, (cls, g) in enumerate(
            zip((FullAttentionManager, SlidingWindowManager), groups)
        )
    ]
    store_mask = sched._reachable_store_block_mask(
        sched.config.kv_group_configs[1], 0, 4, 4, ()
    )
    record = dict(
        extra_retained_tokens=extra,
        older_state=older_state,
        expanded_coverage=expanded_coverage,
        hit_tokens=external,
        async_load=async_load,
        swa_store_mask=store_mask,
    )
    if external:
        for m in managers:
            m.add_local_computed_blocks(req.request_id, [], 0, external)
        for m in managers:
            m.allocate_external_computed_blocks(req.request_id, 0, external)
        blocks = managers[1].req_to_blocks[req.request_id]
        record["allocated_offsets"] = [i for i, b in enumerate(blocks) if not b.is_null]
        try:
            sched.update_state_after_alloc(
                req,
                KVCacheBlocks(tuple(m.req_to_blocks[req.request_id] for m in managers)),
                external,
            )
        except AssertionError as exc:
            record["error"] = str(exc)
        else:
            record["load_jobs"] = len(sched._jobs)
            for job in sched._jobs.values():
                manager.complete_load(job.keys, state.req_context)
    return record


if __name__ == "__main__":
    records = [
        probe(extra, older, expand)
        for expand in (False, True)
        for extra in (0, 3)
        for older in ("missing", "pending", "ready")
    ]
    for r in records:
        if r["extra_retained_tokens"] == 0 or r["older_state"] == "ready":
            assert r["hit_tokens"] == 64 and r["load_jobs"] == 1, r
        elif r["expanded_coverage"]:
            assert r["hit_tokens"] == (0 if r["older_state"] == "missing" else None), r
            assert "error" not in r, r
        else:
            assert r["hit_tokens"] == 64 and "error" in r, r
    output = json.dumps(records, indent=2)
    Path(__file__).with_suffix(".json").write_text(output + "\n")
    print(output)
    print("All 12 diagnostic cases matched their expected outcomes.")
```

</details>

### AI assistance

OpenAI Codex assisted with issue/PR searches, source inspection, the reproducer, execution, and this RFC. This is a design discussion with a verified bookkeeping failure, not a reviewed production patch or a claim of completed model validation.
